File: src/main/java/me/nasukhov/TukitaLearner/bot/bridge/IOResolver.java
package me.nasukhov.TukitaLearner.bot.bridge;

import me.nasukhov.TukitaLearner.bot.io.Channel;
import me.nasukhov.TukitaLearner.bot.io.Output;
import me.nasukhov.TukitaLearner.bot.bridge.tg.Telegram;
import me.nasukhov.TukitaLearner.bot.bridge.tg.TelegramOutput;
import org.springframework.context.ApplicationContext;
import org.springframework.stereotype.Component;

@Component
public class IOResolver {
    public final static String TG_PREFIX = "tg_";

    private final ApplicationContext serviceLocator;

    public IOResolver(ApplicationContext applicationContext) {
        this.serviceLocator = applicationContext;
    }

    public Output resolveFor(Channel channel) {
        if (isTelegramChannel(channel)) {
            var chatId = Long.parseLong(channel.id.substring(TG_PREFIX.length()));

            return new TelegramOutput(chatId, serviceLocator.getBean(Telegram.class));
        }

        throw new RuntimeException("Channel interaction is not supported. Probably some mistake in the code.");
    }

    public static Channel telegramChannel(Long chatId, boolean isPublic) {
        return new Channel(TG_PREFIX + chatId, isPublic);
    }

    private static boolean isTelegramChannel(Channel channel) {
        return channel.id.startsWith(TG_PREFIX);
    }
}

File: src/main/kotlin/me/nasukhov/tukitalearner/bot/bridge/IOResolver.kt
package me.nasukhov.tukitalearner.bot.bridge

import me.nasukhov.tukitalearner.bot.bridge.tg.Telegram
import me.nasukhov.tukitalearner.bot.bridge.tg.TelegramOutput
import me.nasukhov.tukitalearner.bot.io.Channel
import me.nasukhov.tukitalearner.bot.io.Output
import org.springframework.context.ApplicationContext
import org.springframework.stereotype.Component

// TODO injecting entire service locator is rather messy
@Component
class IOResolver(
    private val serviceLocator: ApplicationContext,
) {
    fun resolveFor(channel: Channel): Output {
        require(isTelegramGroup(channel)) {
            "Channel interaction is not supported yet."
        }

        val chatId = channel.id.substring(TG_PREFIX.length).toLong()

        return TelegramOutput(chatId, serviceLocator.getBean(Telegram::class.java))
    }

    companion object {
        const val TG_PREFIX: String = "tg_"

        private fun isTelegramGroup(channel: Channel): Boolean = channel.id.startsWith(TG_PREFIX)
    }
}
